### ISSUE TITLE
geth takes over an hour to start

### ISSUE THREAD

**Issue by Crypto2** [ethereum/go-ethereum#16874]
Hi,

This has been going on for a long time but geth takes longer and longer to start up. Currently using the pre-compiled 1.8.10 on Linux 64-bit. Pretty beefy machine with 128GB of RAM, NVMe storage, 3 GHz Xeon w/ 8 cores / 16 threads.

My best guess is it's because we have a lot of keys since it's for a wallet provider.

Quick example from today showing timestamps between these 2 lines of geth output:

> INFO [06-04|04:26:23] Maximum peer count                       ETH=25 LES=0 total=25
> INFO [06-04|05:38:10] Starting peer-to-peer node               instance=Geth/v1.8.10-stable-eae63c51/linux-amd64/go1.10.

**Comment by jsvisa** [comment]
The time before node starting is the sync time, you can check out https://github.com/ethereum/go-ethereum/issues/16875#issuecomment-394273073 to see more

**Comment by Crypto2** [comment]
This is before syncing ever starts. These are the 1st two lines of geth output.

**Comment by timnugent** [comment]
Same issue here with 1.8.10 - also a powerful machine and previously synced up to block ~5770000. Only a few keys on this one.

**Comment by Crypto2** [comment]
Any update on this? It persists in 1.8.12.

**Comment by Crypto2** [comment]
At least in my case it does seem to be related to having a lot of keys, I renamed my keystore folder and it started up right away as would be expected.

**Comment by Crypto2** [comment]
@jsvisa Yep, just tried this on another machine with the same result.
with keystore = over an hour to start, without keystore = immediate start

**Comment by jsvisa** [comment]
@Crypto2 How much keys under keystore do you have?

**Comment by Crypto2** [comment]
Several million, they should be loaded/scanned in another thread or something?

**Comment by Crypto2** [comment]
Any updates on this?

**Comment by Crypto2** [comment]
Any updates? Still takes hours to start.

**Comment by Crypto2** [comment]
@jsvisa Any updates? I can do the rename as indicated in my last post but it takes nearly a day for the keys to scan in the background. There has to be some way to speed this up?

**Comment by ligi** [comment]
How many keys do you have in your keystore?

**Comment by holiman** [comment]
@Crypto2 could you provide your startup flags ?

**Comment by Crypto2** [comment]
~4.5 million keys

Using: geth --ipcdisable --rpc --rpcapi "eth,net,web3,personal,admin,debug" --cache 16384 --syncmode fast --gcmode archive

**Comment by ligi** [comment]
@Crypto2 thanks for the info - this is most likely the cause of the delay then. Can you elaborate a bit on your use-case? Why do you need 4.5 million keys in your keystore?

**Comment by holiman** [comment]
So, fun fact, although the filename contains the address, geth doesn't rely on that. Instead, geth opens every one of those 4.5 million keys and checks the address within, and places them in a cache, to provide mappings from address to file. 

But I'm also curious why 4.5 million keys would be useful. A crypto exchange might want to consider using custom software to manage accounts in bulk -- geth is mainly trying to be a general usecase node. If there were some simple way of optimizing away that check, that'd be one thing, but that's not really feasible.

**Comment by ligi** [comment]
one thing that could be done for sure:
 - print a warning when the keystore contains too many files so users know this is the culprit

I could imagine optimizations that could be done there - e.g. cache the address->file mapping - but I am not sure this should be done as it really looks like the wrong tool is used for the job in this case.

**Comment by Crypto2** [comment]
Yes it's a hosted wallet. If we knew it was going to be an issue beforehand we would have tried to do something different; it wasn't until years in that it became a problem. I would say there for sure should be some kind of address <-> file mapping since addresses have to be unlocked before use that's the perfect time for it to actually try to load and decrypt the key.

**Comment by Crypto2** [comment]
Any update on this to load the keys faster and/or in the background?

**Comment by adamschmideg** [comment]
This is a feature that could be implemented in Clef but it's far from obvious.

**Comment by Crypto2** [comment]
Any updates on this? Such as:

1. Saving the mapping to a file periodically that can be reloaded on the next start.
2. Scanning the filenames into a list 1st before trying to load them, that way if a key is needed that hasn't been checked yet it can try based on the filename on demand.

**Comment by Crypto2** [comment]
@holiman I ended up making a key server that uses the geth/web3 keystore and will just give me the private key so I can sign TXes in software. I'm not sure if you are interested in making these kinds of improvements to geth/clef but when I was done it will create the address -> file map in about 5 minutes on the first run then 30-60 seconds on follow up runs versus the 9+ hours geth takes now. These were the key optimizations I made:

1. Storing the address -> file map in an LMDB table.
2. When a filename starts with "UTC--20", is 66+ digits long, and ends in "--" followed by 40 hex digits it uses that for the address. Since your keystore folder should be mostly trusted I think that's an OK optimization. It falls back to reading the file if it doesn't match that pattern.
3. I'm not sure if this is possible in Go but on Linux I made it use the getdents64 syscall versus readdir(), this is a massive speedup on it's own.

**Comment by AlexanderYastrebov** [comment]
Hello, I've created https://github.com/ethereum/go-ethereum/pull/32757 to improve startup time significantly.
